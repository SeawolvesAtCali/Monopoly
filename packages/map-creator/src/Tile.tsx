import * as React from 'react';
import { Mode } from './common';
import styles from './App.module.css';
import { Tile, TileType, RoadTile, PropertyTile, StartTile } from '@prisel/monopoly-common';
import { DrawingModeContext } from './DrawingModeHandler';
import { DrawPropertyHandler } from './DrawPropertyHandler';
import { DrawStartHandler } from './DrawStartHandler';
import { DrawRoadHandler } from './DrawRoadHandler';
import { EraseHandler } from './EraseHandler';

interface TileViewProps {
    drawingMode: Mode;
    isMouseDown: boolean;
    row: number;
    column: number;
    onStartPair: DrawingModeContext['onStartPair'];
    onConnectPair: DrawingModeContext['onConnectPair'];
    onClearArrow: DrawingModeContext['onClearArrow'];
}

const drawingModeHandlerMap = {
    [Mode.ERASE_ROAD]: EraseHandler,
    [Mode.ROAD]: DrawRoadHandler,
    [Mode.START]: DrawStartHandler,
    [Mode.PROPERTY]: DrawPropertyHandler,
};

function getTileClass(tileType: TileType): string {
    switch (tileType) {
        case TileType.ROAD:
            return styles.roadTile;
        case TileType.PROPERTY:
            return styles.propertyTile;
        case TileType.START:
            return styles.startTile;
        default:
            return '';
    }
}

export interface TileExport {
    export: () => Tile;
}

export const TileView = React.forwardRef<TileExport, TileViewProps>((props, ref) => {
    const {
        drawingMode,
        isMouseDown,
        onStartPair,
        onConnectPair,
        onClearArrow,
        row,
        column,
    } = props;
    const [selected, setSelected] = React.useState(TileType.UNSPECIFIED);

    const drawingModeContext = React.useMemo<DrawingModeContext>(
        () => ({
            setSelectedMode: setSelected,
            onStartPair,
            onConnectPair,
            onClearArrow,
        }),
        [onStartPair, onConnectPair, onClearArrow],
    );
    const handleMouseOver = React.useCallback(() => {
        if (drawingMode === Mode.UNSPECIFIED || !isMouseDown) {
            return;
        }
        if (drawingModeHandlerMap[drawingMode]) {
            const handler = drawingModeHandlerMap[drawingMode];
            if (handler.disableMouseOver) {
                return;
            }
            handler.onDraw(row, column, selected, drawingModeContext);
        }
    }, [drawingMode, isMouseDown, row, column, selected, drawingModeContext]);
    const handleMouseDown = React.useCallback(() => {
        if (drawingMode === Mode.UNSPECIFIED) {
            return;
        }
        if (drawingModeHandlerMap[drawingMode]) {
            const handler = drawingModeHandlerMap[drawingMode];
            handler.onDraw(row, column, selected, drawingModeContext);
        }
    }, [drawingMode, row, column, selected, drawingModeContext]);

    React.useImperativeHandle(
        ref,
        () => ({
            export: () => {
                switch (selected) {
                    case TileType.START:
                        const startTile: StartTile = {
                            type: TileType.START,
                            pos: { row, col: column },
                            // prev and next will be filled out by BoardView's export
                            next: [],
                            prev: [],
                        };
                        return startTile;
                    case TileType.ROAD:
                        const roadTile: RoadTile = {
                            type: TileType.ROAD,
                            pos: { row, col: column },
                            // prev and next will be filled out by BoardView's export
                            next: [],
                            prev: [],
                        };
                        return roadTile;
                    case TileType.PROPERTY:
                        const propertyTile: PropertyTile = {
                            type: TileType.PROPERTY,
                            pos: { row, col: column },
                            name: 'unnamed property',
                            price: 100,
                            rent: 10,
                        };
                        return propertyTile;
                    default:
                        return {
                            type: TileType.UNSPECIFIED,
                            pos: { row, col: column },
                        };
                }
            },
        }),
        [selected, row, column],
    );
    return (
        <div
            className={`${styles.tile} ${getTileClass(selected)}`}
            onMouseOver={handleMouseOver}
            onMouseDown={handleMouseDown}
        />
    );
});
