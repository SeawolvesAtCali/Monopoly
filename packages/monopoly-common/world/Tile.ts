import { getRand } from '../getRand';
import { createClass } from './createClass';
import { ChancePoolMixinConfig } from './mixins/ChancePoolMixin';
import { HasPropertiesMixinConfig } from './mixins/HasPropertiesMixin';
import { required } from './mixins/MixinConfig';
import { PathMixinConfig } from './mixins/PathMixin';
import { PositionMixinConfig } from './mixins/PositionMixin';
import { StartMixinConfig } from './mixins/StartMixin';

export const TileClass = createClass('tile', [
    required(PositionMixinConfig),
    PathMixinConfig,
    HasPropertiesMixinConfig,
    StartMixinConfig,
    ChancePoolMixinConfig,
]);

export type Tile = InstanceType<typeof TileClass>;

/**
 * generate path not including current node using genNextPathTile funtion.If
 * the function return undefined, generation is stopped.
 */
export const Tiles = {
    genPathWith(
        tile: Tile,
        genNextPathTile: (currentPathTile: Tile, length: number) => Tile | undefined,
    ): Tile[] {
        const path: Tile[] = [];
        let current: Tile = tile;
        while (true) {
            const next = genNextPathTile(current, path.length);
            if (next) {
                path.push(next);
                current = next;
            } else {
                break;
            }
        }
        return path;
    },

    genPath(tile: Tile, steps: number) {
        return Tiles.genPathWith(tile, (current, length) =>
            length === steps
                ? undefined
                : // Choose a random next
                  getRand(current.path?.next ?? [])?.() ?? undefined,
        );
    },

    genPathReverse(tile: Tile, steps: number) {
        return Tiles.genPathWith(tile, (current, length) =>
            length === steps
                ? undefined
                : // Choose a random next
                  getRand(current.path?.prev ?? [])?.() ?? undefined,
        );
    },
};
