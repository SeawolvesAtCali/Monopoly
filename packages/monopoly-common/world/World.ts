import { genId } from '../genId';
import { GameObject, GameObjectClass } from './GameObject';
import { Id } from './Id';
import { Ref, RefIdSymbol } from './Ref';
import { Serialized } from './serialize';

export interface SerializedWorld {
    [key: string]: Array<Serialized<GameObject>>;
}

/**
 * World holds all game objects. Provide easy access to game object through ids
 */
export class World {
    private objectMap: Map<Id, GameObject> = new Map();
    private idMap: Map<string, Set<Id>> = new Map();
    private gameObjectTypeRegistry: Map<string, GameObjectClass<any>> = new Map();
    private serializedTypes = new Set<string>();

    public registerObject<T extends GameObjectClass<any>>(clazz: T, enableSerialization = true) {
        const type = clazz.TYPE;
        this.gameObjectTypeRegistry.set(type, clazz);
        this.idMap.set(type, new Set());
        if (enableSerialization) {
            this.serializedTypes.add(type);
        }
        return this;
    }

    private createId<T extends GameObject>(specifiedId?: Id<T>): Id<T> {
        if (specifiedId && this.objectMap.has(specifiedId)) {
            throw new Error(
                `Cannot create game object of id ${specifiedId} because there is already one existing`,
            );
        }
        if (specifiedId) {
            return specifiedId;
        }
        return genId();
    }

    public get<T extends GameObjectClass<any>>(id: Id<InstanceType<T>>): InstanceType<T> | null {
        const object = this.objectMap.get(id);
        return (object as InstanceType<T>) || null;
    }

    private getRespectiveSet(object: GameObject) {
        const type = object.type;
        const gameObjectClass = this.gameObjectTypeRegistry.get(type);
        if (gameObjectClass && object instanceof gameObjectClass) {
            return this.idMap.get(type) ?? null;
        }
        return null;
    }

    private getAllFromSet<T extends GameObject>(idSet: Set<Id>): T[] {
        return Array.from(idSet)
            .map((id) => this.objectMap.get(id) as T)
            .filter(Boolean);
    }

    public getAll<T extends GameObjectClass<any>>(typeOrClazz: string | T): InstanceType<T>[] {
        const typeString = typeof typeOrClazz === 'string' ? typeOrClazz : typeOrClazz.TYPE;
        const set = this.idMap.get(typeString);
        if (set) {
            return this.getAllFromSet<InstanceType<T>>(set);
        }
        return [];
    }

    /** Create a GameObject of given type, and add to world */
    public create<T extends GameObjectClass<any>>(
        clazz: T,
        id?: Id<InstanceType<T>>,
    ): InstanceType<T> {
        const object = new clazz();
        const objectId = this.createId<InstanceType<T>>(id);
        object.id = objectId;
        this.objectMap.set(objectId, object);
        this.getRespectiveSet(object)?.add(objectId);

        return object;
    }

    public createRef<T extends GameObjectClass<any>>(
        clazz: T,
        id?: Id<InstanceType<T>>,
    ): Ref<InstanceType<T>> {
        return this.getRef(this.create(clazz, id));
    }

    public remove<T extends GameObject>(idOrGameObject: Id | T) {
        let deletedObject;
        if (typeof idOrGameObject === 'string') {
            deletedObject = this.objectMap.get(idOrGameObject);
            this.objectMap.delete(idOrGameObject);
        } else {
            deletedObject = idOrGameObject;
            this.objectMap.delete(idOrGameObject.id);
        }
        if (deletedObject) {
            this.getRespectiveSet(deletedObject)?.delete(deletedObject.id);
        }
    }

    public getRef<T extends GameObjectClass<any>>(
        keyOrObject: Id<InstanceType<T>> | InstanceType<T>,
    ): Ref<InstanceType<T>> {
        const id = typeof keyOrObject === 'string' ? keyOrObject : keyOrObject.id;
        const ref = <Ref<InstanceType<T>>>(() => this.get(id));
        ref[RefIdSymbol] = id;
        return ref;
    }

    public serialize(): SerializedWorld {
        const serialized = {};
        for (const serializedObjectType of Array.from(this.serializedTypes)) {
            Object.assign(serialized, {
                [serializedObjectType]:
                    this.getAll(serializedObjectType).map((object) => object.serialize()) ?? [],
            });
        }
        return serialized;
    }

    /**
     * deserialize into current world make sure this is an empty world,
     * otherwise we might have id collision
     * @param serialized
     */
    public deserialize(serialized: SerializedWorld) {
        for (const key of Object.keys(serialized)) {
            if (this.serializedTypes.has(key) && this.gameObjectTypeRegistry.has(key)) {
                const clazz = this.gameObjectTypeRegistry.get(key);
                if (clazz) {
                    for (const serializedObject of serialized[key]) {
                        clazz.deserialize(serializedObject, this);
                    }
                }
            }
        }
        return this;
    }
}
