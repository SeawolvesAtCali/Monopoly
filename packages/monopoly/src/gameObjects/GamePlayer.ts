import {
    Coordinate,
    deserialize,
    GameObject,
    GamePlayerInfo,
    log,
    Payment,
    Properties,
    Property2,
    PropertyInfo,
    serialize,
    Serialized,
    Tile2,
    Tiles,
    World,
} from '@prisel/monopoly-common';
import { Player, PlayerId } from '@prisel/server';

interface Props {
    id: PlayerId;
    player: Player;
    pathTile: Tile2;
    owning: Property2[];
    cash: number;
    character: number;
    rolled: boolean;
}

function roll(startingNode: Tile2): Tile2[] {
    const steps = 3;
    // const steps = Math.trunc(Math.random() * 6) + 1;
    log.info(`player will move ${steps}`);
    return Tiles.genPath(startingNode, steps);
}

export interface GamePlayer {}
export class GamePlayer extends GameObject {
    public static TYPE: string = 'game_player';
    public get type() {
        return 'game_player';
    }
    public pathTile: Tile2;
    public owning: Property2[];
    public cash: number;
    public rolled: boolean;
    public player: Player;
    public character: number;

    public init(props: Props) {
        this.id = props.id;
        this.pathTile = props.pathTile;
        this.owning = props.owning || [];
        this.cash = props.cash;
        this.rolled = props.rolled;
        this.player = props.player;
        this.character = props.character;
        return this;
    }

    public payRent(owner: GamePlayer, property: PropertyInfo): Payment {
        this.cash = this.cash - property.rent;
        owner.gainMoney(property.rent);
        return {
            from: this.player.getId(),
            to: owner.player.getId(),
            forProperty: property,
            amount: property.rent,
        };
    }

    public gainMoney(amount: number) {
        this.cash += amount;
    }

    public rollAndMove(): Coordinate[] {
        const path = roll(this.pathTile);
        this.rolled = true;
        return this.move(path);
    }

    public move(path: Tile2[]) {
        if (path.length > 0) {
            this.pathTile = path[path.length - 1];
            return path.map((pathTile) => pathTile.position);
        }
        return [];
    }

    public purchaseProperty(property: Property2, nextLevelPropertyInfo: PropertyInfo) {
        this.cash = this.cash - nextLevelPropertyInfo.cost;
        if (property.owner !== this.id) {
            this.owning.push(property);
        }
        if (nextLevelPropertyInfo.isUpgrade) {
            Properties.upgrade(property, nextLevelPropertyInfo.currentLevel);
        } else {
            Properties.purchasedBy(property, this);
        }
    }

    public getGamePlayerInfo(): GamePlayerInfo {
        return {
            money: this.cash,
            player: {
                name: this.player.getName(),
                id: this.player.getId(),
            },
            pos: this.pathTile.position,
            character: this.character,
        };
    }

    public serialize(): Serialized<this> {
        return serialize(this, []);
    }

    public static deserialize(serialized: Serialized<GamePlayer>, world: World) {
        return deserialize(GamePlayer, serialized, [], world);
    }
}

export function create(props: Props): GamePlayer {
    return new GamePlayer().init(props);
}
