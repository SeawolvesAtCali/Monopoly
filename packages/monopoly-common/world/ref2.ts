import { assertExist } from '@prisel/common';
import { serializable } from 'serializr';
import { GameObject } from './GameObject';
import { Id } from './Id';
import { World } from './World';

export class Ref<T extends GameObject> {
    @serializable
    id;
    world: World;

    static of<T extends GameObject>(object: T | Id<T>, world: World): Ref<T> {
        const ref = new Ref<T>(object instanceof GameObject ? object.id : object);
        ref.world = world;
        return ref;
    }

    static forTest<T extends GameObject>(object: T | Id<T>): Ref<T> {
        const ref = new Ref<T>(object instanceof GameObject ? object.id : object);
        return ref;
    }

    constructor(id: Id<T>) {
        this.id = id;
    }

    get(): T {
        return assertExist(this.world, 'world').get(this.id);
    }

    equals(ref: Ref<any>) {
        return ref.id === this.id;
    }
}
