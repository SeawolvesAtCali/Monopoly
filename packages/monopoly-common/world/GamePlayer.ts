import { coordinate, game_player, payment, prompt_purchase_spec, property } from '@prisel/protos';
import { serializable } from 'serializr';
import { GameObject } from './GameObject';
import { Id } from './Id';
import type { Property } from './Property';
import { Ref } from './ref2';
import { listRefSerializable, refSerializable } from './serializeUtil';
import type { Tile } from './Tile';

interface Props {
    id: Id<GamePlayer>;
    player: any;
    pathTile: Tile;
    owning: Property[];
    money: number;
    character: number;
    rolled: boolean;
}

export class GamePlayer extends GameObject {
    public static TYPE = 'game_player';
    public readonly type = 'game_player';
    @serializable
    money = 0;

    @listRefSerializable
    owning: Ref<Property>[] = [];

    @refSerializable
    pathTile: Ref<Tile>;

    @serializable
    rolled = false;

    @serializable
    character = 0;

    player: any;

    forcedRollPoint = 0;

    public init(props: Props) {
        this.id = props.id;
        this.pathTile = this.world.makeRef(props.pathTile);
        this.owning = (props.owning || []).map(this.world.makeRef);
        this.money = props.money;
        this.rolled = props.rolled;
        this.player = props.player;
        this.character = props.character;
        return this;
    }

    public payRent(owner: GamePlayer, property: property.PropertyInfo): payment.Payment {
        this.money = this.money - property.rent;
        owner.gainMoney(property.rent);
        return {
            payer: this.id,
            payee: owner.id,
            forProperty: property,
            amount: property.rent,
        };
    }

    public gainMoney(amount: number) {
        this.money += amount;
    }

    private roll(startingNode: Ref<Tile>): Tile[] {
        if (this.forcedRollPoint) {
            // player will move a fixed ${forcedRollPoint} step
            return startingNode.get().genPath(this.forcedRollPoint);
        }
        const steps = Math.trunc(Math.random() * 6) + 1;

        return startingNode.get().genPath(steps);
    }

    public rollAndMove(): coordinate.Coordinate[] {
        const path = this.roll(this.pathTile);
        this.rolled = true;
        return this.move(path);
    }

    public move(path: Tile[]) {
        if (path.length > 0) {
            this.pathTile = this.world.makeRef(path[path.length - 1]);
            return path.map((pathTile) => pathTile.position);
        }
        return [];
    }

    public purchaseProperty(
        property: Property,
        nextLevelPropertyInfo: prompt_purchase_spec.PromptPurchaseRequest,
    ) {
        if (!nextLevelPropertyInfo.property) {
            return;
        }
        this.money = this.money - nextLevelPropertyInfo.property?.cost;
        if (property.owner?.id !== this.id) {
            this.owning.push(this.world.makeRef(property));
        }
        if (nextLevelPropertyInfo.isUpgrade) {
            property.upgrade(nextLevelPropertyInfo.property?.currentLevel);
        } else {
            property.purchasedBy(this);
        }
    }

    public getGamePlayerInfo(): game_player.GamePlayer {
        return {
            id: this.id,
            money: this.money,
            pos: this.pathTile.get().position,
            character: this.character,
        };
    }
}
