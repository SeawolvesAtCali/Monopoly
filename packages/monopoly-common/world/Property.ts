import { coordinate, payment, prompt_purchase_spec, property } from '@prisel/protos';
import { serializable } from 'serializr';
import { exist } from '../exist';
import { Size } from '../types';
import { GameObject } from './GameObject';
import { GamePlayer } from './GamePlayer';
import { Ref } from './ref2';
import { jsonSerializable, refSerializable } from './serializeUtil';

export class Property extends GameObject {
    static TYPE = 'property';
    readonly type = 'property';

    @jsonSerializable
    anchor: coordinate.Coordinate;

    @jsonSerializable
    size: Size;

    @serializable
    currentLevel = -1;

    @jsonSerializable
    levels: property.PropertyLevel[] = [];

    @serializable
    name = '';

    @refSerializable
    owner?: Ref<GamePlayer>;

    purchasedBy(owner: GamePlayer) {
        this.currentLevel = 0;
        this.owner = Ref.of(owner, this.world);
    }

    upgrade(newLevel: number) {
        this.currentLevel = newLevel;
    }

    purchaseable(): boolean {
        return !exist(this.owner);
    }

    upgradable(requester: GamePlayer): boolean {
        return (
            exist(this.owner) &&
            this.owner.id === requester.id &&
            this.currentLevel < this.levels.length - 1
        );
    }

    investable(requester: GamePlayer): boolean {
        return this.purchaseable() || this.upgradable(requester);
    }

    getWorth() {
        // the worth of a property is the sum of the cost taken to
        // purchase/upgrade the property.
        return this.levels.reduce(
            (prev, cur, level) => (level <= this.currentLevel ? prev + cur.cost : prev),
            0,
        );
    }

    getBasicPropertyInfo(): property.PropertyInfo {
        return {
            name: this.name,
            pos: this.anchor,
            currentLevel: this.currentLevel,
            ...this.levels[this.currentLevel],
        };
    }

    getNextLevel(): property.PropertyInfo {
        return {
            ...this.getBasicPropertyInfo(),
            currentLevel: this.currentLevel + 1,
            ...this.levels[this.currentLevel + 1],
        };
    }

    getPromptPurchaseRequest(
        requester: GamePlayer,
        existingMoney: number,
    ): prompt_purchase_spec.PromptPurchaseRequest | null {
        if (this.currentLevel >= this.levels.length) {
            return null;
        }

        const nextLevel = this.getNextLevel();
        return {
            property: nextLevel,
            levels: this.levels,
            isUpgrade: this.upgradable(requester),
            moneyAfterPurchase: existingMoney - nextLevel.cost,
        };
    }

    getPaymentForRent(property: Property, payer: string, payee: string): payment.Payment | null {
        const basicPropertyInfo = this.getBasicPropertyInfo();
        if (
            basicPropertyInfo.currentLevel < 0 ||
            basicPropertyInfo.currentLevel >= this.levels.length
        ) {
            return null;
        }
        return {
            payer,
            payee,
            amount: this.levels[this.currentLevel].rent,
            forProperty: this.getBasicPropertyInfo(),
        };
    }
}
