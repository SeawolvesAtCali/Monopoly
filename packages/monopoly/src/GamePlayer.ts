import { debug, Player, PlayerId, Request, Messages, Response } from '@prisel/server';
import Property, { toPropertyInfo } from './Property';
import { log } from './logGameObject';
import GameObject, { FlatGameObject, Ref } from './GameObject';
import Game from './Game';
import PathNode from './PathNode';
import { RollResponsePayload, PurchaseResponsePayload, PurchasePayload } from '../common/messages';
import { PropertyInfo, Encounter } from '../common/types';
import { samePos } from './utils';

interface Props {
    id: PlayerId;
    player: Player;
    pathNode: PathNode;
    owning: Property[];
    cash: number;
    character: number;
    rolled: boolean;
}

export interface FlatPlayer extends FlatGameObject {
    position: Ref<PathNode>;
    owning: Array<Ref<Property>>;
    cash: number;
    rolled: boolean;
}

function roll(startingNode: PathNode): PathNode[] {
    const steps = Math.trunc(Math.random() * 6) + 1;
    debug(`player will move ${steps}`);
    return startingNode.genPath(steps);
}

export class GamePlayer extends GameObject {
    public pathNode: PathNode;
    public owning: Property[];
    public cash: number;
    public rolled: boolean;
    public player: Player;
    public character: number;

    constructor(props: Props) {
        super();
        this.id = props.id;
        this.pathNode = props.pathNode;
        this.owning = props.owning || [];
        this.cash = props.cash;
        this.rolled = props.rolled;
        this.player = props.player;
        this.character = props.character;
    }

    @log
    private payRent(owner: GamePlayer, property: Property): PropertyInfo {
        this.cash = this.cash - property.rent;
        owner.gainMoney(property.rent);
        return toPropertyInfo(property);
    }

    @log
    public gainMoney(amount: number) {
        this.cash += amount;
    }

    @log
    public roll(game: Game, packet: Request): Response<RollResponsePayload> {
        if (this.rolled) {
            return Messages.getFailureFor(packet, `Player ${this.id} already rolled`);
        }

        const path = roll(this.pathNode);
        this.rolled = true;
        this.pathNode = path[path.length - 1];
        const { properties } = this.pathNode;
        const encounters: Encounter[] = [];
        if (properties.length > 0) {
            const payRentProperties: PropertyInfo[] = [];
            const propertiesForPurchase: PropertyInfo[] = [];
            // check for rent payment first
            for (const property of properties) {
                if (property.owner && property.owner.id !== this.id) {
                    payRentProperties.push(this.payRent(property.owner, property));
                }
                if (!property.owner) {
                    propertiesForPurchase.push(toPropertyInfo(property));
                }
            }
            if (payRentProperties.length > 0) {
                encounters.push({
                    payRent: {
                        properties: payRentProperties,
                        remainingMoney: this.cash,
                    },
                });
            }
            if (propertiesForPurchase.length > 0) {
                encounters.push({
                    newPropertyForPurchase: {
                        properties: propertiesForPurchase,
                    },
                });
            }
            // TODO when player doesn't have enough cash, declare bankrupcy
        }

        return Messages.getSuccessFor<RollResponsePayload>(packet, {
            path: path.map((pathNode) => pathNode.tile.pos),
            encounters,
        });
    }

    @log
    public purchase(
        game: Game,
        request: Request<PurchasePayload>,
    ): Response<PurchaseResponsePayload> {
        const { properties } = this.pathNode;
        const propertyToPurchase = properties.find((property) =>
            samePos(property.pos, request.payload.propertyPos),
        );

        if (!propertyToPurchase) {
            return Messages.getFailureFor(
                request,
                `no property at location ${request.payload.propertyPos}`,
            );
        }
        if (propertyToPurchase.owner) {
            return Messages.getFailureFor(request, 'property is already owned');
        }
        if (propertyToPurchase.cost > this.cash) {
            return Messages.getFailureFor(
                request,
                `not enough cash to purchase, current cash ${this.cash}, property price ${propertyToPurchase.cost}`,
            );
        }

        this.cash = this.cash - propertyToPurchase.cost;
        propertyToPurchase.setOwner(this);
        this.owning.push(propertyToPurchase);
        return Messages.getSuccessFor<PurchaseResponsePayload>(request, {
            property: toPropertyInfo(propertyToPurchase),
            remainingMoney: this.cash,
        });
    }

    @log
    public endTurn(game: Game, packet: Request): Response {
        if (this.rolled) {
            this.rolled = false;
            return Messages.getSuccessFor(packet);
        }
        // if not rolled, cannot end turn
        return Messages.getFailureFor(packet, 'Not rolled yet');
    }

    public flat(): FlatPlayer {
        return {
            id: this.id,
            position: this.ref(this.pathNode),
            owning: this.owning.map(this.ref),
            cash: this.cash,
            rolled: this.rolled,
        };
    }
}

export function create(props: Props): GamePlayer {
    return new GamePlayer(props);
}
