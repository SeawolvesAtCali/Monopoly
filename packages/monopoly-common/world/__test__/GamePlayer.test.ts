import { GamePlayer } from '../GamePlayer';
import { Property } from '../Property';
import { Ref } from '../ref2';
import { Tile } from '../Tile';
import { World } from '../World';

describe('GamePlayer', () => {
    test('serialize', () => {
        const player = new GamePlayer();
        player.character = 1;
        player.id = '123';
        player.money = 100;
        player.owning = [Ref.forTest<Property>('property1'), Ref.forTest<Property>('property2')];
        player.pathTile = Ref.forTest<Tile>('tile1');
        player.rolled = true;
        expect(player.serialize()).toMatchObject({
            id: '123',
            type: 'game_player',
            data: {
                money: 100,
                owning: [{ id: 'property1' }, { id: 'property2' }],
                pathTile: { id: 'tile1' },
                rolled: true,
                character: 1,
            },
        });
    });

    test('deserialize', () => {
        const player = new GamePlayer();
        player.character = 1;
        player.id = '123';
        player.money = 100;
        player.owning = [Ref.forTest<Property>('property1'), Ref.forTest<Property>('property2')];
        player.pathTile = Ref.forTest<Tile>('tile1');
        player.rolled = true;
        const world = new World();
        const deserialized = GamePlayer.deserialize(player.serialize(), world);
        expect(deserialized).toBeInstanceOf(GamePlayer);
        expect(deserialized.id).toBe('123');
        expect(deserialized.money).toBe(100);
        expect(deserialized.owning).toHaveLength(2);
        expect(deserialized.owning[0].equals(Ref.forTest('property1'))).toBe(true);
        expect(deserialized.owning[1].equals(Ref.forTest('property2'))).toBe(true);
        expect(deserialized.pathTile.equals(Ref.forTest('tile1'))).toBe(true);
        expect(deserialized.rolled).toBe(true);
        expect(deserialized.world).toBe(world);
    });
});
