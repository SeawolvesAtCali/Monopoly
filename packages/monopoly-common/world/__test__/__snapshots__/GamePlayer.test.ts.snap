// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePlayer serialize 1`] = `
Object {
  "data": Object {
    "character": 1,
    "collectibles": Array [
      Object {
        "description": "activate when entering jail",
        "id": "mocked_id_1",
        "name": "get out of jail free",
        "type": 1,
      },
    ],
    "money": 100,
    "owning": Array [
      "property1",
      "property2",
    ],
    "pathTile": "tile1",
    "rolled": true,
  },
  "id": "123",
  "type": "game_player",
}
`;
